src/coprocessor/metrics.rs

```diff
@@ -153,3 +183,249 @@ pub fn tls_collect_read_flow(region_id: u64, statistics: &Statistics) {
         flow_stats.add(&statistics.data.flow_stats);
     });
 }
+
+pub fn tls_collect_qps(region_id: u64, peer: &metapb::Peer, start_key: &[u8], end_key: &[u8]) {
+    TLS_COP_METRICS.with(|m| {
+        let mut m = m.borrow_mut();
+        m.hub.add(region_id, peer, start_key, end_key);
+    });
+}
+
+const QPS_THRESHOLD: u64 = 100;
+const DETECT_TIMES: u32 = 10;
+const TOP_N: u32 = 10;
+const DETECT_INTERVAL: Duration = Duration::from_secs(1);
+
+pub struct Sample {
+    pub key: Vec<u8>,
+    pub left: i32,
+    pub contained: i32,
+    pub right: i32,
+}
+
+fn build_sample(key: &[u8]) -> Sample {
+    Sample {
+        key: key.to_owned(),
+        left: 0,
+        contained: 0,
+        right: 0,
+    }
+}
+
+pub struct KeyRange {
+    pub start_key: Vec<u8>,
+    pub end_key: Vec<u8>,
+    pub qps: u64,
+}
+
+fn build_key_range(start_key: &[u8], end_key: &[u8]) -> KeyRange {
+    KeyRange {
+        start_key: start_key.to_owned(),
+        end_key: end_key.to_owned(),
+        qps: 0,
+    }
+}
+
+pub struct Recorder {
+    pub samples: Vec<Sample>,
+    pub times: u32,
+    pub count: u64,
+    pub create_time: SystemTime,
+}
+
+fn build_recorder() -> Recorder {
+    Recorder {
+        samples: vec![],
+        times: 0,
+        count: 0,
+        create_time: SystemTime::now(),
+    }
+}
+
+impl Recorder {
+    fn record(&mut self, key_ranges: &[KeyRange]) {
+        self.times += 1;
+        for key_range in key_ranges.iter() {
+            self.count += 1;
+            if self.samples.len() < 20 {
+                self.samples.push(build_sample(&key_range.start_key));
+            } else {
+                let i = rand::thread_rng().gen_range(0, self.count) as usize;
+                if i < 20 {
+                    self.samples[i] = build_sample(&key_range.start_key);
+                }
+            }
+            self.sample(key_range);
+        }
+    }
+
+    fn sample(&mut self, key_range: &KeyRange) {
+        for mut sample in self.samples.iter_mut() {
+            if sample.key.cmp(&key_range.start_key) == Ordering::Less {
+                sample.left += 1;
+            } else if !key_range.end_key.is_empty()
+                && sample.key.cmp(&key_range.end_key) == Ordering::Greater
+            {
+                sample.right += 1;
+            } else {
+                sample.contained += 1;
+            }
+        }
+    }
+
+    fn split_key(&self) -> Vec<u8> {
+        info!("start-split";"times"=>self.times,"De"=>DETECT_TIMES);
+        if self.times < DETECT_TIMES {
+            return vec![];
+        }
+        let mut best_index: i32 = -1;
+        let mut best_score = 2.0;
```

Comment:
Since there's only one score, best_score can be initialized as `1.0`.

Reply:
Good idea